tracer/src/Datadog.Trace/AppSec/Security.cs

```diff
@@ -121,28 +158,90 @@ private static void AnnotateSpan(Span span)
             }
         }
 
+        private static void LogMatchesIfDebugEnabled(WafMatch[] results, bool blocked)
+        {
+            if (Log.IsEnabled(LogEventLevel.Debug))
+            {
+                for (var i = 0; i < results.Length; i++)
+                {
+                    var result = results[i];
+                    if (blocked)
+                    {
+                        Log.Debug("Blocking current transaction (rule: {RuleId})", result.Rule);
+                    }
+                    else
+                    {
+                        Log.Debug("Detecting an attack from rule {RuleId}", result.Rule);
+                    }
+                }
+            }
+        }
+
+        private static void AddHeaderTags(Span span, IHeadersCollection headers, Dictionary<string, string> headersToCollect)
```

Comment:
To avoid boxing the `HeadersCollectionAdapter`, I think we can change `GetRequestHeaders()` and `GetResponseHeaders()` to return `HeadersCollectionAdapter` (instead of `IHeadersCollection`), and this signature to:
```suggestion
        private static void AddHeaderTags<THeaders>(Span span, THeaders headers, Dictionary<string, string> headersToCollect)
            where THeaders : IHeadersCollection
```